File: demo_backend/src/main/java/com/andrei/demo/service/ProjectService.java
package com.andrei.demo.service;

import com.andrei.demo.model.Project;
import com.andrei.demo.model.ProjectCreateDTO;
import com.andrei.demo.repository.ProjectRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@AllArgsConstructor
public class ProjectService {
    private final ProjectRepository projectRepository;

    public List<Project> getProjects() {
        return projectRepository.findAll();
    }

    public Project getProjectById(UUID id) {
        return projectRepository.findById(id)
                .orElseThrow(() -> new IllegalStateException("Project not found"));
    }

    public Project addProject(ProjectCreateDTO dto) {
        Project project = new Project();
        project.setTitle(dto.getTitle());   // <-- use title, not name
        return projectRepository.save(project);
    }

    public Project updateProject(UUID id, ProjectCreateDTO dto) {
        Project project = getProjectById(id);
        project.setTitle(dto.getTitle());   // <-- use title, not name
        return projectRepository.save(project);
    }

    public void deleteProject(UUID id) {
        projectRepository.deleteById(id);
    }
}

File: demo-app/src/app/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Project {
  id: string;
  title: string;
}

export interface ProjectCreateDTO {
  title: string;
}

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly apiUrl = 'http://localhost:8082/project';

  constructor(private http: HttpClient) {}

  getProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(this.apiUrl);
  }

  getProjectById(id: string): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }

  addProject(dto: ProjectCreateDTO): Observable<Project> {
    return this.http.post<Project>(this.apiUrl, dto);
  }

  updateProject(id: string, dto: ProjectCreateDTO): Observable<Project> {
    return this.http.put<Project>(`${this.apiUrl}/${id}`, dto);
  }

  deleteProject(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
